src/beignet/_newton_raphson.py

```diff
@@ -0,0 +1,172 @@
+        y_target_batch = y_target_batch.expand(B)
+    if y_target_batch.shape[0] != B:
+        raise ValueError("y_target_batch must have shape (B,) or be a scalar.")
+
+    if max_update_step <= 0:
+        raise ValueError("max_update_step must be positive.")
+
+    # Initialize x
+    x: Tensor
+    if isinstance(x_init, torch.Tensor):
+        x = x_init.clone().to(device, dtype)
+        if x.shape[0] != B:
+            raise ValueError(
+                f"x_init tensor shape {x.shape} incompatible with batch size {B}"
+            )
+    else:
+        # Ensure x is created with the correct dtype
+        x = torch.full_like(y_target_batch, float(x_init), dtype=dtype)
+
+    # Ensure initial x respects bounds
+    if min_bound is not None:
+        x = torch.clamp(x, min=min_bound)
+    if max_bound is not None:
+        x = torch.clamp(x, max=max_bound)
+
+    # Track active elements (those not yet converged)
+    active: Tensor = torch.ones_like(y_target_batch, dtype=torch.bool)
+
+    for i in range(max_iter):
+        if not torch.any(active):
+            break
+
+        # --- Operate only on active elements ---
+        active_indices: Tensor = torch.where(active)[0]
+        # Ensure active_indices isn't empty before proceeding
+        if active_indices.numel() == 0:
+            break
+        x_active: Tensor = x[active]
+        params_active: Tensor = params_batch[active]
+        y_target_active: Tensor = y_target_batch[active]
+
+        # --- Calculate func and f_prime ---
+        try:
+            y_val: Tensor
+            aux_data: Any
+            y_prime_val: Tensor
+            y_val, aux_data = func(x_active, params_active)
+            y_prime_val = f_prime(x_active, params_active, aux_data)
+        except RuntimeError as e:
+            warnings.warn(
+                f"RuntimeError at (iter {i}): {e}. Check inputs/x values.",
+                UserWarning,
+                stacklevel=2,
+            )
+            y_prime_val = torch.ones_like(x_active)
+
+        # --- Calculate Error ---
+        g_val: Tensor = y_val - y_target_active
+
+        # --- Convergence Check (Absolute Tolerance) ---
+        converged_now: Tensor = torch.abs(g_val) < tol
+        active[active_indices[converged_now]] = False  # Deactivate converged elements
+
+        if not torch.any(active):  # Break if all converged this iteration
+            break
+
+        # --- Update Step (for non-converged elements) ---
+        update_mask: Tensor = ~converged_now
+        if not torch.any(update_mask):
+            continue  # Skip update if none left
+
+        active_indices_update: Tensor = active_indices[update_mask]
+        if active_indices_update.numel() == 0:
+            continue
+
+        g_val_update: Tensor = g_val[update_mask]
+        y_prime_val_update: Tensor = y_prime_val[update_mask]
+
+        # Newton Step: update = error / derivative
+        raw_update: Tensor = g_val_update / (y_prime_val_update + 1e-12)
```

Comment:
You’re performing a massive update then clamping. It’ll, at best, converge slowly and, at worst, stall.

Reply:
def does slow convergence, was the only thing I'd found so far that prevented oscillation on my test cases.